File: generate_epg.js
const fs = require('fs');
const path = require('path');
const { create } = require('xmlbuilder2');

const sitesDir = path.join(__dirname, 'sites');

// إنشاء الجذر للـ XML
const root = create({ version: '1.0', encoding: 'UTF-8' }).ele('tv');

// دالة لتحليل JS واستخراج القنوات
function extractChannelsFromJS(filePath) {
    const content = fs.readFileSync(filePath, 'utf-8');
    // نفترض أن الملف يحتوي على متغير JS باسم channels
    const regex = /const channels\s*=\s*([\s\S]*?);/m;
    const match = content.match(regex);
    if (!match) return [];
    let channels = [];
    try {
        channels = eval(match[1]); // تحويل النص إلى كائن JS
    } catch (e) {
        console.error('خطأ في تحليل JS:', filePath, e);
    }
    return channels;
}

// المرور على كل مجلد فرعي
fs.readdirSync(sitesDir).forEach(siteFolder => {
    const folderPath = path.join(sitesDir, siteFolder);
    if (fs.statSync(folderPath).isDirectory()) {
        const jsFiles = fs.readdirSync(folderPath).filter(f => f.endsWith('.js'));
        jsFiles.forEach(jsFile => {
            const jsPath = path.join(folderPath, jsFile);
            const channels = extractChannelsFromJS(jsPath);
            channels.forEach(ch => {
                const chEle = root.ele('channel', { id: ch.name });
                chEle.ele('display-name').txt(ch.name);
                chEle.ele('icon').txt(ch.icon || '');
                chEle.ele('url').txt(ch.url || '');
                // إضافة البرامج إذا موجودة
                if (ch.programs) {
                    ch.programs.forEach(p => {
                        root.ele('programme', {
                            start: p.start

File: generate_epg.py
import os
import xml.etree.ElementTree as ET

# إنشاء الجذر للـ XML
root = ET.Element("tv")

# مسار مجلد المواقع
sites_dir = "sites"

# المرور عبر كل ملف في مجلد sites
for filename in os.listdir(sites_dir):
    if filename.endswith(".xml"):
        file_path = os.path.join(sites_dir, filename)
        tree = ET.parse(file_path)
        channel = tree.getroot()

        # استخراج البيانات من القناة
        for ch in channel.findall("channel"):
            display_name = ch.find("display-name").text
            icon_url = ch.find("icon").text if ch.find("icon") is not None else ""
            url = ch.find("url").text if ch.find("url") is not None else ""

            # إنشاء عنصر جديد في XML
            tv_channel = ET.SubElement(root, "channel", id=display_name)
            ET.SubElement(tv_channel, "display-name").text = display_name
            ET.SubElement(tv_channel, "icon").text = icon_url
            ET.SubElement(tv_channel, "url").text = url

# إنشاء الشجرة وكتابة الملف
tree = ET.ElementTree(root)
tree.write("epg.xml", encoding="UTF-8", xml_declaration=True)
